wen-restart/src/last_voted_fork_slots_aggregate.rs

```diff
@@ -3,69 +3,108 @@ use {
+}
+
 pub(crate) struct LastVotedForkSlotsAggregate {
-    root_slot: Slot,
-    repair_threshold: f64,
-    // TODO(wen): using local root's EpochStakes, need to fix if crossing Epoch boundary.
-    epoch_stakes: EpochStakes,
+    // Map each peer pubkey to the epoch of its last vote.
+    node_to_last_vote_epoch_map: HashMap<Pubkey, Epoch>,
+    epoch_info_vec: Vec<LastVotedForkSlotsEpochInfo>,
     last_voted_fork_slots: HashMap<Pubkey, RestartLastVotedForkSlots>,
-    slots_stake_map: HashMap<Slot, u64>,
-    active_peers: HashSet<Pubkey>,
-    slots_to_repair: HashSet<Slot>,
     my_pubkey: Pubkey,
+    repair_threshold: f64,
+    root_bank: Arc<Bank>,
+    slots_stake_map: HashMap<Slot, u64>,
+    slots_to_repair: BTreeSet<Slot>,
 }
 
 #[derive(Clone, Debug, PartialEq)]
 pub struct LastVotedForkSlotsFinalResult {
     pub slots_stake_map: HashMap<Slot, u64>,
-    pub total_active_stake: u64,
+    pub epoch_info_vec: Vec<LastVotedForkSlotsEpochInfo>,
 }
 
 impl LastVotedForkSlotsAggregate {
     pub(crate) fn new(
-        root_slot: Slot,
+        root_bank: Arc<Bank>,
         repair_threshold: f64,
-        epoch_stakes: &EpochStakes,
         last_voted_fork_slots: &Vec<Slot>,
         my_pubkey: &Pubkey,
     ) -> Self {
-        let mut active_peers = HashSet::new();
-        let sender_stake = Self::validator_stake(epoch_stakes, my_pubkey);
-        active_peers.insert(*my_pubkey);
         let mut slots_stake_map = HashMap::new();
+        let root_slot = root_bank.slot();
+        let root_epoch = root_bank.epoch();
         for slot in last_voted_fork_slots {
             if slot >= &root_slot {
-                slots_stake_map.insert(*slot, sender_stake);
+                if let Some(sender_stake) = root_bank.epoch_node_id_to_stake(root_epoch, my_pubkey)
+                {
+                    slots_stake_map.insert(*slot, sender_stake);
+                } else {
+                    warn!(
+                        "The root bank {} does not have the stake for slot {}",
+                        root_slot, slot
+                    );
+                }
             }
         }
+        let last_vote_epoch = root_bank
+            .get_epoch_and_slot_index(last_voted_fork_slots[0])
+            .0;
+        let mut node_to_last_vote_epoch_map = HashMap::new();
+        node_to_last_vote_epoch_map.insert(*my_pubkey, last_vote_epoch);
+        // We would only consider slots in root_epoch and the next epoch.
+        let epoch_info_vec: Vec<LastVotedForkSlotsEpochInfo> = (root_epoch
+            ..root_epoch
+                .checked_add(2)
+                .expect("root_epoch should not be so big"))
+            .map(|epoch| LastVotedForkSlotsEpochInfo {
+                epoch,
+                total_stake: root_bank
+                    .epoch_total_stake(epoch)
+                    .expect("epoch stake not found"),
+                voted_percent: 0.0,
+                voted_for_this_epoch_percent: 0.0,
```

Comment:
Should this voted percent/voted_for_this_epoch_percent be updated for our own validator's vote?

Reply:
Done.